sandbox/samples/cpp/distancefieldobject/distancefieldobject.cpp

```diff
@@ -0,0 +1,315 @@
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+// appleseed.renderer headers.
+#include "renderer/api/object.h"
+#include "renderer/api/project.h"
+#include "renderer/api/rendering.h"
+#include "renderer/api/scene.h"
+#include "renderer/api/types.h"
+
+// todo: fix.
+#include "renderer/kernel/shading/shadingray.h"
+
+// appleseed.foundation headers.
+#include "foundation/math/hash.h"
+#include "foundation/math/intersection/rayaabb.h"
+#include "foundation/math/ray.h"
+#include "foundation/math/rng/distribution.h"
+#include "foundation/math/rng/xoroshiro128plus.h"
+#include "foundation/math/scalar.h"
+#include "foundation/math/vector.h"
+#include "foundation/platform/types.h"
+#include "foundation/utility/api/specializedapiarrays.h"
+#include "foundation/utility/containers/dictionary.h"
+#include "foundation/utility/job/iabortswitch.h"
+#include "foundation/utility/casts.h"
+#include "foundation/utility/searchpaths.h"
+#include "foundation/utility/string.h"
+
+// appleseed.main headers.
+#include "main/dllvisibility.h"
+
+// Standard headers.
+#include <algorithm>
+#include <cmath>
+#include <cstddef>
+
+namespace asf = foundation;
+namespace asr = renderer;
+
+namespace
+{
+    //
+    // An object whose surface is defined by a signed distance field.
+    //
+
+    const char* Model = "distance_field_object";
+
+    class DistanceFieldObject
+      : public asr::ProceduralObject
+    {
+      public:
+        // Constructor.
+        DistanceFieldObject(
+            const char*                 name,
+            const asr::ParamArray&      params)
+          : asr::ProceduralObject(name, params)
+          , m_lazy_region_kit(&m_region_kit)
+        {
+        }
+
+        // Delete this instance.
+        void release() override
+        {
+            delete this;
+        }
+
+        // Return a string identifying this object model.
+        const char* get_model() const override
+        {
+            return Model;
+        }
+
+        // Compute the local space bounding box of the object over the shutter interval.
+        asr::GAABB3 compute_local_bbox() const override
+        {
+            return asr::GAABB3(asr::GVector3(-1.0f), asr::GVector3(1.0f));
+        }
+
+        // Return the region kit of the object.
+        asf::Lazy<asr::RegionKit>& get_region_kit() override
+        {
+            return m_lazy_region_kit;
+        }
+
+        // Access materials slots.
+        size_t get_material_slot_count() const override
+        {
+            return 1;
+        }
```

Comment:
Newline?

Reply:
I've left the newline out since these two methods go together under the banner "Access materials slots."